internal/target/dtrack_target.go

```diff
@@ -0,0 +1,112 @@
+package target
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	dtrack "github.com/nscuro/dtrack-client"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+
+	"github.com/ckotzbauer/sbom-operator/internal"
+)
+
+type DependencaTrackTarget struct {
+	baseUrl        string
+	apiKey         string
+	imageToProject map[string]uuid.UUID
+}
+
+func NewDependencaTrackTarget() *DependencaTrackTarget {
+	baseUrl := viper.GetString(internal.ConfigKeyDependencaTrackBaseUrl)
+	apiKey := viper.GetString(internal.ConfigKeyDependencaTrackApiKey)
+	return &DependencaTrackTarget{
+		baseUrl: baseUrl,
+		apiKey:  apiKey,
+	}
+}
+
+func (g *DependencaTrackTarget) ValidateConfig() error {
+	if g.baseUrl == "" {
+		return fmt.Errorf("%s is empty", internal.ConfigKeyDependencaTrackBaseUrl)
+	}
+	if g.apiKey == "" {
+		return fmt.Errorf("%s is empty", internal.ConfigKeyDependencaTrackApiKey)
+	}
+	return nil
+}
+
+func (g *DependencaTrackTarget) Initialize() {
+	client, _ := dtrack.NewClient(g.baseUrl, dtrack.WithAPIKey(g.apiKey))
+	projectsPage, err := client.Project.GetAll(context.TODO(), dtrack.PageOptions{
+		PageNumber: 1,
+		PageSize:   10,
+	})
+	if err != nil {
+		logrus.Errorf("Could not fetch projects: %v", err)
+	}
+
+	g.imageToProject = make(map[string]uuid.UUID)
+	for _, project := range projectsPage.Projects {
+		for _, property := range project.Properties {
+			if property.Name == "image-name" {
+				g.imageToProject[property.Value] = project.UUID
+			}
+		}
+	}
+}
+
+func (g *DependencaTrackTarget) ProcessSbom(imageID, sbom string) {
+	if sbom == "" {
+		logrus.Infof("Empty SBOM - skip image (image=%s)", imageID)
+		return
+	}
+
+	client, _ := dtrack.NewClient(g.baseUrl, dtrack.WithAPIKey(g.apiKey))
+	logrus.Infof("Sending SBOM to Dependency Track (image=%s)", imageID)
+
+	if !strings.ContainsRune(imageID, '@') {
+		logrus.Warnf("Image id %s does not contain an @sha256", imageID)
+		return
+	}
+	imageSplit := strings.Split(imageID, "@")
+	imageName := imageSplit[0]
+
+	projectId := g.imageToProject[imageName]
+	if projectId == uuid.Nil {
+		project, err := client.Project.Create(context.TODO(),
+			dtrack.Project{
+				Active:     true,
+				Classifier: "APPLICATION",
+				Name:       imageName,
+				Properties: []dtrack.ProjectProperty{
+					{Name: "image-name", Group: "container", Value: imageName, Type: "STRING"},
+				},
```

Comment:
The SBOM is uploaded to a named-project here (without version). What happens when there are multiple versions of the same image in the cluster? Would DT only know the last uploaded SBOM then (for whatever version)?

Reply:
The question is what do we use as as version? 
* The image tag? - Not all images have a tag. Also, the `image` from the pod is not available in `ProcessSbom`
* The image hash? - Maybe too many versions get created.

Reply:
The URN for an image deployed to K8s has 3 components: registry hostname, image name, tag.  The tag is essentially a version.   

Some registries allow the tag to be updated, and there are special tags like `latest` which mean that sometimes an image could be processed that has the same name and tag but different packages in it.  In this case you would just upload it to the same project and let DT handle updating it's internal view of that project.

If an image doesn't have a tag, then you should use the value `latest` as that is the default alias applied to an empty tag value:  https://docs.docker.com/engine/reference/commandline/tag/

Reply:
If the `ProcessSbom` function needs more infos, then the target-interface can be changed at any time.
Currently the `ProcessSbom` gets the imageID from the pod (e.g. `ghcr.io/ckotzbauer/sbom-operator@sha256:face2223025ee1a616b90f3a04fea920e7785436174531eed31255262edd318b`). This image is currently referenced  by the tag `ghcr.io/ckotzbauer/sbom-operator:0.3.1`. I don't know which version would be better for DT.

Reply:
I know all this - so you are suggesting to use the tag? @ckotzbauer if we want the tag we have to change the signature of `ProcessSbom` to include the image tag - currently this only contains the `imageID` which does not include the tag.

Reply:
Yes, it would be possible to extend the `ContainerImage` struct by a tag-field and update the `LoadImageInfos` func accordingly: https://github.com/ckotzbauer/sbom-operator/blob/e731e895fe579794bbf784b4e9b237209d331647/internal/kubernetes/kubernetes.go#L83

Then the `ProcessSbom` could either receive another parameter with the tag or the whole `ContainerImage` struct, I think the latter would be the best solution.

Reply:
> @pmckeown Some registries allow the tag to be updated, and there are special tags like latest which mean that sometimes an image could be processed that has the same name and tag but different packages in it. 

☝ This. Tags are mutable. You can have multiple images with the same tag, but different contents and thus different digests. You'll see this often when the underlying OS layer of an image is updated, while the application stays the same. It's totally possible to have multiple versions (read: contents/digests) of an image running across your cluster(s). 

> In this case you would just upload it to the same project and let DT handle updating it's internal view of that project.

Considering Syft inventorizes not only the application, but the whole image, I think using the digest as version would make more sense. For better readability, you could use the `tag@sha256:<DIGEST>` notation and additionally store tag and digest as project properties, if desired. When using just the tag, you're only seeing the contents of the image with that tag that was scanned most recently. Which does not necessarily represent what's present in your entire cluster.

Reply:
Changed the implementation, now it uses:
* the image repository (registry+image name) as the project name
* the image tag as the version (with latest fallback)

I have also changed the code so that it always just calls the SBOM upload with `auto-create=true` (as suggested by @pmckeown in #5). This makes the code a lot simpler. Just tested that on our clusters and it works as expected.

@ckotzbauer what do you think?